algorithms/rainbow/agent.py

```diff
@@ -0,0 +1,158 @@
+            documentation.
+            Source: https://pytorch.org/docs/stable/torch.html#torch.load
+            """
+            self.online_net.load_state_dict(torch.load(args.model, map_location='cpu'))
+        self.online_net.train()
+
+        self.target_net = RainbowDQN(args, self.action_space).to(device=args.device)
+        self.update_target_net()
+        self.target_net.train()
+        for param in self.target_net.parameters():
+            param.requires_grad = False
+
+        self.optimiser = optim.Adam(self.online_net.parameters(), lr=args.lr, eps=args.adam_eps)
+
+    def reset_noise(self):
+        """ Resets noisy weights in all linear layers (of online net only) """
+        self.online_net.reset_noise()
+
+    def act(self, state):
+        """ Acts based on single state (no batch) """
+        with torch.no_grad():
+            return (self.online_net(state.unsqueeze(0)) * self.support).sum(2).argmax(1).item()
+
+    def act_e_greedy(self, state, epsilon=0.001):
+        """
+        Acts with an ε-greedy policy (used for evaluation only)
+        High ε can reduce evaluation scores drastically
+        """
+        return np.random.randint(0, self.action_space.n) if np.random.random() < epsilon \
+            else self.act(state)
+
+    def learn(self, mem):
+        """
+        Executes 1 gradient descent step sampling batch_size transitions from the memory
+        """
+        # Sample transitions
+        idxs, states, actions, returns, next_states, nonterminals, weights = \
+          mem.sample(self.batch_size)
+
+        # Calculate current state probabilities (online network noise already sampled)
+        log_ps = self.online_net(states, log=True)  # Log probabilities log p(s_t, ·; θonline)
+        log_ps_a = log_ps[range(self.batch_size), actions]  # log p(s_t, a_t; θonline)
+
+        with torch.no_grad():
+            # TODO: more detailed explanation of this process
+            # Calculate nth next state probabilities
+            # Probabilities p(s_t+n, ·; θonline)
+            pns = self.online_net(next_states)
+            # Distribution d_t+n = (z, p(s_t+n, ·; θonline))
+            dns = self.support.expand_as(pns) * pns
+            # Perform argmax action selection using online network:
+            # argmax_a[(z, p(s_t+n, a; θonline))]
+            argmax_indices_ns = dns.sum(2).argmax(1)
+            # Sample new target net noise
+            self.target_net.reset_noise()
+            # Probabilities p(s_t+n, ·; θtarget)
+            pns = self.target_net(next_states)
+            # Double-Q probabilities p(s_t+n, argmax_a[(z, p(s_t+n, a; θonline))]; θtarget)
+            pns_a = pns[range(self.batch_size), argmax_indices_ns]
+
+            # Compute Tz (Bellman operator T applied to z)
+            # Tz = R^n + (γ^n)z (accounting for terminal states)
+            Tz = returns.unsqueeze(1) + nonterminals * (self.discount ** self.n) \
```

Comment:
help me understand self.n and n-step q-learning

Reply:
You just found a bug sir, but I will answer your question regardless before I fix it.
In N-step Q-Learning (from Sutton) we make N steps and collect all those returns, then use Q to estimate the expected return, having Q_target = R1 + R2 + R3 + Q(s,a), analogous to TD n-step. Now you probably see the formula much simpler

Reply:
I've always generally known this but none of the code seems to suggest that we actually store R1 + R2 + R3. Or is it in the returns itself somehow? I'm trying to do a ctrl+f on self.n... self.n is maybe a bad name too but not too bad.
I'm still confused a bit about it being called nonterminals as well. Maybe we'll take this in slack

Reply:
The key to understanding that is mem.sample(self.batch_size) in that script. The returns output from the mem.sample function are already summed for "truncated n-steps", i.e. if multi_step parameter is set to 3, 3 consecutive steps from the memory are retrieved and their rewards summed before appending that value to returns. You do that operation until you obtain a tensor with size self.batch_size. That's why you see here only the last operation. Non-terminals is a boolean that checks if you did get enough transitions or actually got to terminal state before or right at the end (after a terminal state there is no step in theory). Hope this is clearer now

Reply:
It is clearer after I looked at the _get_sample_from_segment() function from memory.py. 
You should maybe say that in a comment (don't make it two lines).
`# look in _get_sample_from_segment() from memory.py for more details`